pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java

```diff
@@ -580,17 +589,26 @@ private List<Map<String, Object>> processValidDocIdMetadata(String tableNameWith
           LOGGER.warn(msg);
           continue;
         }
-        MutableRoaringBitmap validDocIds =
-            indexSegment.getValidDocIds() != null ? indexSegment.getValidDocIds().getMutableRoaringBitmap() : null;
-        if (validDocIds == null) {
+
+        // Adopt the same logic as the query execution to get the valid doc ids. 'FilterPlanNode.run()'
+        // If the queryableDocId is available (upsert delete is enabled), we read the valid doc ids from it.
+        // Otherwise, we read the valid doc ids.
+        MutableRoaringBitmap validDocIdSnapshot = null;
+        if (indexSegment.getQueryableDocIds() != null) {
+          validDocIdSnapshot = indexSegment.getQueryableDocIds().getMutableRoaringBitmap();
+        } else if (indexSegment.getValidDocIds() != null) {
+          validDocIdSnapshot = indexSegment.getValidDocIds().getMutableRoaringBitmap();
+        }
+
+        if (validDocIdSnapshot == null) {
```

Comment:
```suggestion
        } else {
```
Same as above, this logic is that `validDocIdSnapshot` must be assigned one of the two values and if both are null than an exception is thrown.  I think this structure makes the logic more explicit.

Reply:
fixed